### ISSUE TITLE
when converting to stanfit object naming scheme is inconsistent

### ISSUE THREAD

**Issue by yonicd** [stan-dev/cmdstanr#447]
**Describe the bug**
Inconsistent naming scheme for `stanfit@sim$samples[[1]]` and `stanfit@sim$fnames_oi` that departs from naming scheme in `{rstan}` creating revdep package problems. eg `{tidybayes}` cc @mjskay

**To Reproduce**

``` r
library(magrittr)

library(cmdstanr)
#> This is cmdstanr version 0.3.0.9000
#> - Online documentation and vignettes at mc-stan.org/cmdstanr
#> - CmdStan path set to: /Users/yonis/.cmdstanr/cmdstan-2.26.0
#> - Use set_cmdstan_path() to change the path
library(rstan)
#> Loading required package: StanHeaders
#> Loading required package: ggplot2
#> rstan (Version 2.21.2, GitRev: 2e1f913d3ca3)
#> For execution on a local, multicore CPU with excess RAM we recommend calling
#> options(mc.cores = parallel::detectCores()).
#> To avoid recompilation of unchanged Stan programs, we recommend calling
#> rstan_options(auto_write = TRUE)
#> 
#> Attaching package: 'rstan'
#> The following object is masked from 'package:magrittr':
#> 
#>     extract

library(tidybayes)

fit <- cmdstanr_example(
  example = "logistic", 
  method = "sample", 
  save_warmup = TRUE
)
#> Compiling Stan program...

stanfit <- rstan::read_stan_csv(fit$output_files())

names(stanfit@sim$samples[[1]])
#>   [1] "alpha"       "beta.1"      "beta.2"      "beta.3"      "log_lik.1"  
#>   [6] "log_lik.2"   "log_lik.3"   "log_lik.4"   "log_lik.5"   "log_lik.6"  
#>  [11] "log_lik.7"   "log_lik.8"   "log_lik.9"   "log_lik.10"  "log_lik.11" 
#>  [16] "log_lik.12"  "log_lik.13"  "log_lik.14"  "log_lik.15"  "log_lik.16" 
#>  [21] "log_lik.17"  "log_lik.18"  "log_lik.19"  "log_lik.20"  "log_lik.21" 
#>  [26] "log_lik.22"  "log_lik.23"  "log_lik.24"  "log_lik.25"  "log_lik.26" 
#>  [31] "log_lik.27"  "log_lik.28"  "log_lik.29"  "log_lik.30"  "log_lik.31" 
#>  [36] "log_lik.32"  "log_lik.33"  "log_lik.34"  "log_lik.35"  "log_lik.36" 
#>  [41] "log_lik.37"  "log_lik.38"  "log_lik.39"  "log_lik.40"  "log_lik.41" 
#>  [46] "log_lik.42"  "log_lik.43"  "log_lik.44"  "log_lik.45"  "log_lik.46" 
#>  [51] "log_lik.47"  "log_lik.48"  "log_lik.49"  "log_lik.50"  "log_lik.51" 
#>  [56] "log_lik.52"  "log_lik.53"  "log_lik.54"  "log_lik.55"  "log_lik.56" 
#>  [61] "log_lik.57"  "log_lik.58"  "log_lik.59"  "log_lik.60"  "log_lik.61" 
#>  [66] "log_lik.62"  "log_lik.63"  "log_lik.64"  "log_lik.65"  "log_lik.66" 
#>  [71] "log_lik.67"  "log_lik.68"  "log_lik.69"  "log_lik.70"  "log_lik.71" 
#>  [76] "log_lik.72"  "log_lik.73"  "log_lik.74"  "log_lik.75"  "log_lik.76" 
#>  [81] "log_lik.77"  "log_lik.78"  "log_lik.79"  "log_lik.80"  "log_lik.81" 
#>  [86] "log_lik.82"  "log_lik.83"  "log_lik.84"  "log_lik.85"  "log_lik.86" 
#>  [91] "log_lik.87"  "log_lik.88"  "log_lik.89"  "log_lik.90"  "log_lik.91" 
#>  [96] "log_lik.92"  "log_lik.93"  "log_lik.94"  "log_lik.95"  "log_lik.96" 
#> [101] "log_lik.97"  "log_lik.98"  "log_lik.99"  "log_lik.100" "lp__"
stanfit@sim$fnames_oi
#>   [1] "alpha"        "beta[1]"      "beta[2]"      "beta[3]"      "log_lik[1]"  
#>   [6] "log_lik[2]"   "log_lik[3]"   "log_lik[4]"   "log_lik[5]"   "log_lik[6]"  
#>  [11] "log_lik[7]"   "log_lik[8]"   "log_lik[9]"   "log_lik[10]"  "log_lik[11]" 
#>  [16] "log_lik[12]"  "log_lik[13]"  "log_lik[14]"  "log_lik[15]"  "log_lik[16]" 
#>  [21] "log_lik[17]"  "log_lik[18]"  "log_lik[19]"  "log_lik[20]"  "log_lik[21]" 
#>  [26] "log_lik[22]"  "log_lik[23]"  "log_lik[24]"  "log_lik[25]"  "log_lik[26]" 
#>  [31] "log_lik[27]"  "log_lik[28]"  "log_lik[29]"  "log_lik[30]"  "log_lik[31]" 
#>  [36] "log_lik[32]"  "log_lik[33]"  "log_lik[34]"  "log_lik[35]"  "log_lik[36]" 
#>  [41] "log_lik[37]"  "log_lik[38]"  "log_lik[39]"  "log_lik[40]"  "log_lik[41]" 
#>  [46] "log_lik[42]"  "log_lik[43]"  "log_lik[44]"  "log_lik[45]"  "log_lik[46]" 
#>  [51] "log_lik[47]"  "log_lik[48]"  "log_lik[49]"  "log_lik[50]"  "log_lik[51]" 
#>  [56] "log_lik[52]"  "log_lik[53]"  "log_lik[54]"  "log_lik[55]"  "log_lik[56]" 
#>  [61] "log_lik[57]"  "log_lik[58]"  "log_lik[59]"  "log_lik[60]"  "log_lik[61]" 
#>  [66] "log_lik[62]"  "log_lik[63]"  "log_lik[64]"  "log_lik[65]"  "log_lik[66]" 
#>  [71] "log_lik[67]"  "log_lik[68]"  "log_lik[69]"  "log_lik[70]"  "log_lik[71]" 
#>  [76] "log_lik[72]"  "log_lik[73]"  "log_lik[74]"  "log_lik[75]"  "log_lik[76]" 
#>  [81] "log_lik[77]"  "log_lik[78]"  "log_lik[79]"  "log_lik[80]"  "log_lik[81]" 
#>  [86] "log_lik[82]"  "log_lik[83]"  "log_lik[84]"  "log_lik[85]"  "log_lik[86]" 
#>  [91] "log_lik[87]"  "log_lik[88]"  "log_lik[89]"  "log_lik[90]"  "log_lik[91]" 
#>  [96] "log_lik[92]"  "log_lik[93]"  "log_lik[94]"  "log_lik[95]"  "log_lik[96]" 
#> [101] "log_lik[97]"  "log_lik[98]"  "log_lik[99]"  "log_lik[100]" "lp__"

stanfit%>%
  tidybayes::gather_draws(beta[i])
#> Error in spread_draws_long_(draws, variable_names, dimension_names, regex = regex, : No variables found matching spec: beta[i]

fix_names <- function(obj){
  obj@sim$samples <- lapply(obj@sim$samples,function(x){
    names(x) <- obj@sim$fnames_oi
    x
  })
  obj
}

stanfit_fix <- stanfit%>%
  fix_names()

stanfit_fix%>%
  tidybayes::gather_draws(beta[i])
#> # A tibble: 12,000 x 6
#> # Groups:   i, .variable [3]
#>        i .chain .iteration .draw .variable .value
#>    <int>  <int>      <int> <int> <chr>      <dbl>
#>  1     1      1          1     1 beta      -0.866
#>  2     1      1          2     2 beta      -0.712
#>  3     1      1          3     3 beta      -0.349
#>  4     1      1          4     4 beta      -0.118
#>  5     1      1          5     5 beta      -1.34 
#>  6     1      1          6     6 beta      -0.668
#>  7     1      1          7     7 beta      -0.670
#>  8     1      1          8     8 beta      -0.757
#>  9     1      1          9     9 beta      -1.05 
#> 10     1      1         10    10 beta      -0.980
#> # … with 11,990 more rows
```

<sup>Created on 2021-02-07 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessionInfo()
#> R version 3.6.3 (2020-02-29)
#> Platform: x86_64-apple-darwin15.6.0 (64-bit)
#> Running under: macOS Catalina 10.15.7
#> 
#> Matrix products: default
#> BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
#> LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
#> 
#> locale:
#> [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
#> 
#> attached base packages:
#> [1] stats     graphics  grDevices datasets  utils     methods   base     
#> 
#> other attached packages:
#> [1] tidybayes_2.3.1      rstan_2.21.2         ggplot2_3.3.2       
#> [4] StanHeaders_2.21.0-7 cmdstanr_0.3.0.9000  magrittr_2.0.1      
#> 
#> loaded via a namespace (and not attached):
#>  [1] Rcpp_1.0.5           lattice_0.20-38      tidyr_1.1.2         
#>  [4] prettyunits_1.1.1    ps_1.5.0             utf8_1.1.4          
#>  [7] assertthat_0.2.1     digest_0.6.27        V8_3.4.0            
#> [10] plyr_1.8.6           R6_2.5.0             backports_1.2.1     
#> [13] reprex_1.0.0         stats4_3.6.3         evaluate_0.14       
#> [16] coda_0.19-4          highr_0.8            pillar_1.4.7        
#> [19] rlang_0.4.10         curl_4.3             data.table_1.13.6   
#> [22] rstudioapi_0.13      callr_3.5.1          checkmate_2.0.0     
#> [25] rmarkdown_2.6        stringr_1.4.0        loo_2.4.1           
#> [28] munsell_0.5.0        compiler_3.6.3       xfun_0.20           
#> [31] pkgconfig_2.0.3      pkgbuild_1.1.0       htmltools_0.5.0     
#> [34] tidyselect_1.1.0     tibble_3.0.5         gridExtra_2.3       
#> [37] arrayhelpers_1.1-0   codetools_0.2-16     matrixStats_0.58.0  
#> [40] fansi_0.4.2          crayon_1.3.4         dplyr_1.0.3         
#> [43] withr_2.3.0          distributional_0.2.2 ggdist_2.4.0        
#> [46] grid_3.6.3           jsonlite_1.7.2       gtable_0.3.0        
#> [49] lifecycle_0.2.0      posterior_0.1.3      scales_1.1.0        
#> [52] RcppParallel_5.0.2   cli_2.2.0            stringi_1.5.3       
#> [55] farver_2.0.3         renv_0.12.5          fs_1.5.0            
#> [58] ellipsis_0.3.1       generics_0.1.0       vctrs_0.3.6         
#> [61] tools_3.6.3          forcats_0.5.1        svUnit_1.0.3        
#> [64] glue_1.4.2           purrr_0.3.4          abind_1.4-5         
#> [67] processx_3.4.5       parallel_3.6.3       yaml_2.2.1          
#> [70] inline_0.3.17        colorspace_1.4-1     knitr_1.31
```

</details>

**Expected behavior**
`stanfit@sim$samples[[1]]` follow rstan consistent naming scheme.

**Operating system**

```r
#> R version 3.6.3 (2020-02-29)
#> Platform: x86_64-apple-darwin15.6.0 (64-bit)
#> Running under: macOS Catalina 10.15.7
```

**CmdStanR version number**

```r
#> This is cmdstanr version 0.3.0.9000
#> - Online documentation and vignettes at mc-stan.org/cmdstanr
#> - CmdStan path set to: /Users/yonis/.cmdstanr/cmdstan-2.26.0
```

**Additional context**

**Comment by rok-cesnovar** [comment]
Hi, thanks for the bug report. I do however think that if this is a bug, its a bug with `rstan::read_stan_csv` as everything below and including the line 
```
stanfit <- rstan::read_stan_csv(fit$output_files())
```
is not using cmdstanr anymore, its just rstan/tidybayes at/after that point.

I think you can just do the following and rely on rstan anymore:
```
posterior::as_draws_df(fit$draws()) %>%
  tidybayes::gather_draws(beta[i])
```

There might be other solutions and I think @mjskay is also adding direct support (not sure on the state of that) for cmdstanr or rather the posterior package.

**Comment by yonicd** [comment]
Thanks. 

The error though comes from the headers of the csv created by cmdstan which have the . In the names. rstan is just reading the csv.

the posterior solution could work but is inefficient in large fit objects where all the params are converted into a wide dataframe, so if you need a targeted query it isn’t well suited.

**Comment by rok-cesnovar** [comment]
> The error though comes from the headers of the csv created by cmdstan which have the . In the names. rstan is just reading the csv.

This is true. But rstan could fix the names for you as cmdstanr does. 

CmdStanR does not have the control over the cmdstan CSV format and the CSV columns naming scheme for non-scalar parameters is not likely to be changed anytime soon. This has been the standard naming practice since the CSV output was introduced. Changing the CSV naming format would break backwards compatibility and cause oter 

> the posterior solution could work but is inefficient in large fit objects where all the params are converted into a wide dataframe, so if you need a targeted query it isn’t well suited.

Yeah, its not the most efficient. But on the other hand, rstan::read_stan_csv is much less efficient compared to $draws().

There are probably other solutions, this was just the first one I found. The most efficient one will definitely be when tidybayes supports

```
fit %>%
      tidybayes::gather_draws(beta[i])
```

I would advise also looking at https://github.com/mjskay/tidybayes/issues/267 

Bottom line is that I dont think there is anything we can do to help fix this on the cmdstanr side.I think waiting for official tidybayes support is the way forward here.

**Comment by yonicd** [comment]
Thanks for the explanation, I’ll keep an eye on the tidybayes side

**Comment by mjskay** [comment]
@yonicd apart from the issue @rok-cesnovar mentioned you may also be interested in this issue on the tidybayes side: https://github.com/mjskay/tidybayes/issues/132

Basically, I would like to make the name parsing a bit more flexible at some point to support other naming schemes like the `"."` one.
